net/relayd/src/opnsense/mvc/app/controllers/OPNsense/Relayd/Api/SettingsController.php

```diff
@@ -0,0 +1,294 @@
+ *    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *    POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+namespace OPNsense\Relayd\Api;
+
+use \OPNsense\Base\ApiMutableModelControllerBase;
+use \OPNsense\Core\Config;
+use \OPNsense\Relayd\Relayd;
+use \OPNsense\Base\UIModelGrid;
+
+/**
+ * Class SettingsController
+ * @package OPNsense\Relayd
+ */
+class SettingsController extends ApiMutableModelControllerBase
+{
+
+    static protected $internalModelName = 'relayd';
+    static protected $internalModelClass = '\OPNsense\Relayd\Relayd';
+
+    /**
+     * list with valid model node types
+     */
+    private $nodeTypes = array('general', 'host', 'tablecheck', 'table', 'protocol', 'virtualserver');
+
+    /**
+     * query relayd settings
+     * @param $nodeType
+     * @param $uuid
+     * @return result array
+     */
+    public function getAction($nodeType = null, $uuid = null)
+    {
+        $result = array("result" => "failed");
+        if ($this->request->isGet() && $nodeType != null) {
+            $this->validateNodeType($nodeType);
+            $mdlRelayd = new Relayd();
+            if ($nodeType == 'general') {
+                $node = $mdlRelayd->getNodeByReference($nodeType);
+            } else {
+                if($uuid != null) {
+                    $node = $mdlRelayd->getNodeByReference($nodeType . '.' . $uuid);
+                } else {
+                    $node = $mdlRelayd->$nodeType->Add();
+                }
+            }
+            if ($node != null) {
+                $result[$nodeType] = $node->getNodes();
+                $result["result"] = "ok";
+            }
+        }
+        return $result;
+    }
+
+    /**
+     * set relayd properties
+     * @param $nodeType
+     * @param $uuid
+     * @param $action set or add node
+     * @return status array
+     */
+    public function setAction($nodeType = null, $uuid = null)
+    {
+        $result = array("result" => "failed", "validations" => array());
+        if ($this->request->isPost() && $this->request->hasPost("relayd") && $nodeType != null) {
+            $this->validateNodeType($nodeType);
+            $mdlRelayd = new Relayd();
+            if ($nodeType == 'general') {
+                $node = $mdlRelayd->getNodeByReference($nodeType);
+            } else {
+                if($uuid != null) {
+                    $node = $mdlRelayd->getNodeByReference($nodeType . '.' . $uuid);
+                } else {
+                    $node = $mdlRelayd->$nodeType->Add();
+                }
+            }
+            if ($node != null) {
+                $relaydInfo = $this->request->getPost("relayd");
+
+                // perform plugin specific validations
+                if ($nodeType == 'virtualserver') {
```

Comment:
does this need to be so complicated? or is this trying to do all in one controller? a lot of plugins use separate models and controllers to reduce complexity